src/main/java/teammates/ui/webapi/action/SearchLinksAction.java

```diff
@@ -0,0 +1,175 @@
+
+        populateCourseIds(students, instructors);
+        populateCourseIdToInstituteMap();
+
+        List<SearchLinksStudentData> studentsBundle = getStudentsBundle(students);
+        List<SearchLinksInstructorData> instructorsBundle = getInstructorsBundle(instructors);
+        SearchLinksData links = new SearchLinksData(studentsBundle, instructorsBundle);
+
+        return new JsonResult(links);
+    }
+
+    private void populateCourseIds(List<StudentAttributes> students, List<InstructorAttributes> instructors) {
+        for (StudentAttributes student : students) {
+            if (student.course != null) {
+                courseIds.add(student.course);
+            }
+        }
+        for (InstructorAttributes instructor : instructors) {
+            if (instructor.courseId != null) {
+                courseIds.add(instructor.courseId);
+            }
+        }
+    }
+
+    private void populateCourseIdToInstituteMap() {
+        for (String courseId : courseIds) {
+            String instructorForCourseGoogleId = findAvailableInstructorGoogleIdForCourse(courseId);
+            AccountAttributes account = logic.getAccount(instructorForCourseGoogleId);
+            if (account == null) {
+                continue;
+            }
+
+            String institute = StringHelper.isEmpty(account.institute) ? "None" : account.institute;
+            courseIdToInstituteMap.put(courseId, institute);
+        }
+    }
+
+    /**
+     * Finds the googleId of a registered instructor with co-owner privileges.
+     * If there is no such instructor, finds the googleId of a registered
+     * instructor with the privilege to modify instructors.
+     *
+     * @param courseId
+     *            the ID of the course
+     * @return the googleId of a suitable instructor if found, otherwise an
+     *         empty string
+     */
+    private String findAvailableInstructorGoogleIdForCourse(String courseId) {
+        List<InstructorAttributes> instructorList = logic.getInstructorsForCourse(courseId);
+
+        for (InstructorAttributes instructor : instructorList) {
+            if (instructor.isRegistered()
+                    && (instructor.hasCoownerPrivileges()
+                    || instructor.isAllowedForPrivilege(Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_INSTRUCTOR))) {
+                courseIdToInstructorGoogleIdMap.put(courseId, instructor.googleId);
+                return instructor.googleId;
+            }
+
+        }
+
+        return "";
+    }
+
+    private List<SearchLinksStudentData> getStudentsBundle(List<StudentAttributes> students) {
+        List<SearchLinksStudentData> studentsBundle = new ArrayList<>();
+
+        for (StudentAttributes student : students) {
+            if (student.email == null) {
+                continue;
+            }
+
+            SearchLinksStudentData studentData = new SearchLinksStudentData();
+
+            if (student.email != null && student.course != null
```

Comment:
Can we remove this `student.email` condition?

Reply:
Yea my bad it should go too